Documentation/design-docs/dllmap.md

```diff
@@ -0,0 +1,291 @@
+
+Right now, CoreCLR has no good way to handle the differences between platforms when it comes to p/invoking native libraries. Here is an example usage of DllImport:
+```c#
+// Use DllImport to import the Win32 MessageBox function.
+[DllImport("user32.dll", CharSet = CharSet.Unicode)]
+public  static  extern  int  MessageBox(IntPtr  hWnd, String text, String caption, uint type);
+```
+This import works with Windows, but it doesn’t work with any other OS. If run e.g. on Linux, a DllNotFoundException will be thrown, which means that a DLL specified in a DLL import cannot be found.
+
+### Mono’s Dllmap
+
+Mono already provides a feature that addresses the problem of cross-platform p/invoke support. Mono’s [Dllmap](http://www.mono-project.com/docs/advanced/pinvoke/dllmap/) 
+enables to configure p/invoke signatures at runtime. By providing an XML configuration file, 
+user can define a custom mapping between OS-specific library names and methods. 
+Thanks to that, even if a library defined in DllImport is incompatible with an OS that is currently running the application, 
+a correct unmanaged method can be called (if it exists for this OS).
+
+In Mono Dllmap feature custom mapping can be tightly specified based on the OS name, CPU name and a wordsize.
+
+This simple Mono example maps references to the cygwin1.dll library to the libc.so.6 file:
+```xml
+<configuration>
+    <dllmap  dll="cygwin1.dll" target="libc.so.6"/>
+</configuration>
+```
+
+Mono dllmap logic is implemented in [metadata/mono-config.c](https://github.com/mono/mono/blob/master/mono/metadata/mono-config.c) 
+and [metadata/loader.c](https://github.com/mono/mono/blob/master/mono/metadata/loader.c) files.
+
+## Expectations
+
+### .NET Core Dllmap
+
+.NET Core Dllmap’s purpose is to deliver a cross-platform support for p/invoke mechanism in .NET. With Dllmap user will be able to control interop methods by defining custom mapping between OS-specific dlls and methods.
+
+Dllmap will allow making changes in both library names and target method names (entrypoints). Changing entrypoint name will be optional (it will remain unchanged by default).
+
+Target platforms for this feature are: Windows, Linux and OS X.
+
+There will be a diagnostic mechanism available to monitor dllmap related issues.
+
+### Interaction with Dllmap
+
+#### Mono compatibility
+Dllmap should be easy to use. It’s possible to achieve this easily by keeping Mono-compatible style of XML mapping configuration file. 
+It’s described [here](http://www.mono-project.com/docs/advanced/pinvoke/dllmap/). 
+Thanks to compatibility with Mono, users will be able to migrate from Mono’s Dllmap to .NET Core cross-platform applications.
+
+#### Flexibility
+For users, who want to manage their cross-platform dll imports in their own way, dll-load specific events will be exposed. 
```

Comment:
Like above, please call out that this _is_ the future and Dllmap is purely for compat. New users should chose this method.